apps/omg_watcher_rpc/lib/web/validators/merge_constraints.ex

```diff
@@ -0,0 +1,67 @@
+# Copyright 2019-2020 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.WatcherRPC.Web.Validator.MergeConstraints do
+  @moduledoc """
+  Validates `/transaction.merge` parameters
+  """
+
+  alias OMG.Utils.HttpRPC.Validator.Base
+  alias OMG.WatcherRPC.Web.Validator.Helpers
+
+  import OMG.Utils.HttpRPC.Validator.Base
+
+  require OMG.State.Transaction.Payment
+
+  defp get_constraints(params) do
+    case params do
+      %{"address" => _, "currency" => _} ->
+        [
+          {"address", [:address], :address},
+          {"currency", [:currency], :currency}
+        ]
+
+      %{"utxo_positions" => _} ->
+        [{"utxo_positions", [min_length: 2, list: &to_utxo_pos/1], :utxo_positions}]
+
+      _ ->
+        {:error, :operation_bad_request}
+    end
+  end
+
+  @doc """
+  Parses and validates request body for `/transaction.merge`
+  """
+  @spec parse(map()) :: {:ok, map()} | Base.validation_error_t()
```

Comment:
This all thing may be cleaner written this way:
```
def parse(params) do
  with {:ok, constraints} <- get_constraints(params),
       {:ok, result} <- Helpers.validate_constraints(params, constraints) do
    {:ok, result}
  end
end

defp get_constraints(params) do 
  case params do
    %{"address" => _, "currency" => _} ->
      {:ok, [
        {"address", [:address], :address},
        {"currency", [:currency], :currency}
      ]}

    %{"utxo_positions" => _} ->
      {:ok, [{"utxo_positions", [min_length: 2, list: &to_utxo_pos/1], :utxo_positions}]}

    _ ->
      {:error, :operation_bad_request}
  end
end

defp to_utxo_pos(utxo_pos_string) do
  expect(%{"utxo_pos" => utxo_pos_string}, "utxo_pos", :non_neg_integer)
end
```

Note that I removed the `Map.new(result)` for consistency with other endpoints that return a Keyword list when validating params. If you use this, you'll need to update your controllers.

Reply:
Hi Med. I reviewed whether we can use a keyword list. Basically the parameters are passed onto a multi-clause function where the pattern matching happens as follows:

``` elixir
  def merge(%{utxo_positions: utxo_positions}) do
  ... 
  def merge(%{address: address, currency: currency}) do
  ... 
```
Though we could refactor this to use keyword lists, the logic becomes more awkward IMO. So I would opt to keep using a map.

Reply:
Can we maybe convert it to a map later on in the code (when we actually need a map), so all our constraint validator returns the same types?

Reply:
Refactored a47e76d with a `merge` / `handle_merge` approach... let me know